[RNMobile] Limit inner blocks nesting depth to avoid call stack size exceeded crash

fluiddot opened the pull request:
**Related PRs:**
* https://github.com/wordpress-mobile/gutenberg-mobile/pull/6191
* https://github.com/wordpress-mobile/WordPress-Android/pull/19163
* https://github.com/wordpress-mobile/WordPress-iOS/pull/21553

<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Limits `InnerBlock` component to render blocks that are deeply nested, in order to prevent a crash related to exceeding the maximum call stack trace set by the Hermes engine.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Fixes https://github.com/wordpress-mobile/WordPress-Android/issues/18750.
Fixes https://github.com/wordpress-mobile/gutenberg-mobile/issues/6123

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

In the `InnerBlocks` component we check the number of parent blocks associated with the block, this determines the nesting depth level. If it exceeds a threshold defined [here](https://github.com/WordPress/gutenberg/blob/e028e9ada1c06cb8e31e9b848e492ad37422ba65/packages/block-editor/src/components/inner-blocks/warning-max-depth-exceeded.native.js#L78-L82), we render a warning message box instead of rendering the inner blocks.

https://github.com/WordPress/gutenberg/blob/e028e9ada1c06cb8e31e9b848e492ad37422ba65/packages/block-editor/src/components/inner-blocks/index.native.js#L130-L139

Additionally, I extracted the UI elements used in [the "missing" block](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/missing/edit.native.js) (i.e. the block used to render unsupported blocks) into a separate component (`BlockFallbackWebVersion`). In a separate PR, I'll update that block to use this new component.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

#### The warning IS NOT displayed when NOT exceeding the threshold
1. In the web version of the editor, create a post with the following HTML:

<details><summary>Deeply nested block structure (8 levels)</summary>

```
<!-- wp:heading -->
<h2 class="wp-block-heading">Group block</h2>
<!-- /wp:heading -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#cb6767"}}} -->
<p class="has-background" style="background-color:#cb6767">Level 1 </p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#b89759"}}} -->
<p class="has-background" style="background-color:#b89759">Level 2</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#816531"}}} -->
<p class="has-background" style="background-color:#816531">Level 3</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#86c64d"}}} -->
<p class="has-background" style="background-color:#86c64d">Level 4</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#58b873"}}} -->
<p class="has-background" style="background-color:#58b873">Level 5</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#58b8b8"}}} -->
<p class="has-background" style="background-color:#58b8b8">Level 6</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5880b8"}}} -->
<p class="has-background" style="background-color:#5880b8">Level 7</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5870b8"}}} -->
<p class="has-background" style="background-color:#5870b8">Level 8</p>
<!-- /wp:paragraph --></div>
<!-- /wp:group --></div>
<!-- /wp:group -->

<!-- wp:paragraph -->
<p></p>
<!-- /wp:paragraph --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group -->
```

</details> 

2. Open the post in the app.
3. Scroll down the content.
4. Observe that all blocks are rendered and that no warning message box is displayed.

#### The warning IS displayed when exceeding the limit
1. In the web version of the editor, create a post with the following HTML in a WPCOM site:

<details><summary>Deeply nested block structure (20 levels)</summary>

```
<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#cb6767"}}} -->
<p class="has-background" style="background-color:#cb6767">Level 1 </p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#b89759"}}} -->
<p class="has-background" style="background-color:#b89759">Level 2</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#816531"}}} -->
<p class="has-background" style="background-color:#816531">Level 3</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#86c64d"}}} -->
<p class="has-background" style="background-color:#86c64d">Level 4</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#58b873"}}} -->
<p class="has-background" style="background-color:#58b873">Level 5</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#58b8b8"}}} -->
<p class="has-background" style="background-color:#58b8b8">Level 6</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5880b8"}}} -->
<p class="has-background" style="background-color:#5880b8">Level 7</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5870b8"}}} -->
<p class="has-background" style="background-color:#5870b8">Level 8</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5877b8"}}} -->
<p class="has-background" style="background-color:#5877b8">Level 9</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5873b8"}}} -->
<p class="has-background" style="background-color:#5873b8">Level 10</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5873b8"}}} -->
<p class="has-background" style="background-color:#5873b8">Level 11</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5873b8"}}} -->
<p class="has-background" style="background-color:#5873b8">Level 12</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5873b8"}}} -->
<p class="has-background" style="background-color:#5873b8">Level 13</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5873b8"}}} -->
<p class="has-background" style="background-color:#5873b8">Level 14</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5873b8"}}} -->
<p class="has-background" style="background-color:#5873b8">Level 15</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5873b8"}}} -->
<p class="has-background" style="background-color:#5873b8">Level 16</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5873b8"}}} -->
<p class="has-background" style="background-color:#5873b8">Level 17</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5873b8"}}} -->
<p class="has-background" style="background-color:#5873b8">Level 18</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5873b8"}}} -->
<p class="has-background" style="background-color:#5873b8">Level 19</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5873b8"}}} -->
<p class="has-background" style="background-color:#5873b8">Level 20</p>
<!-- /wp:paragraph --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group -->

<!-- wp:paragraph -->
<p></p>
<!-- /wp:paragraph --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group -->
```

</details>

2. Open the post in the app.
3. Observe that no crash is produced.
4. Scroll down the content.
5. Observe that starting at level 10, a warning message box is displayed within the block list.
6. Tap on the warning twice.
7. Observe that a bottom sheet is displayed explaining the problem.
8. Observe that two options are presented, one for editing using the web editor and another one for ungrouping the block.
9. Tap on the "Edit using web editor" option.
10. Observe that UBE (Unsupported Block Editor) opens with the content that is not rendered in the native editor.
11. Modify the content and tap on "Continue".
12. Preview the post and observe the post reflects recent modifications.
13. Back in the editor tap on the warning message box again.
14. Tap on the "Ungroup block" option.
15. Observe that a new block appears in the block list. This block should be the one that wasn't previously rendered due to being below level 10.
16. Observe that the warning message box is still present.
17. Tap on the warning message box and then the "Dismiss" option.
18. Observe that the bottom sheet closes.
19. Save the post.
20. Observe that no crash is produced.

#### Option for editing using the web editor is not present in self-hosted sites without Jetpack
1. In the web version of the editor, create a post with the following HTML in a self-hosted site without Jetpack:

<details><summary>Deeply nested block structure (20 levels)</summary>

```
<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#cb6767"}}} -->
<p class="has-background" style="background-color:#cb6767">Level 1 </p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#b89759"}}} -->
<p class="has-background" style="background-color:#b89759">Level 2</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#816531"}}} -->
<p class="has-background" style="background-color:#816531">Level 3</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#86c64d"}}} -->
<p class="has-background" style="background-color:#86c64d">Level 4</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#58b873"}}} -->
<p class="has-background" style="background-color:#58b873">Level 5</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#58b8b8"}}} -->
<p class="has-background" style="background-color:#58b8b8">Level 6</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5880b8"}}} -->
<p class="has-background" style="background-color:#5880b8">Level 7</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5870b8"}}} -->
<p class="has-background" style="background-color:#5870b8">Level 8</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5877b8"}}} -->
<p class="has-background" style="background-color:#5877b8">Level 9</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5873b8"}}} -->
<p class="has-background" style="background-color:#5873b8">Level 10</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5873b8"}}} -->
<p class="has-background" style="background-color:#5873b8">Level 11</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5873b8"}}} -->
<p class="has-background" style="background-color:#5873b8">Level 12</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5873b8"}}} -->
<p class="has-background" style="background-color:#5873b8">Level 13</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5873b8"}}} -->
<p class="has-background" style="background-color:#5873b8">Level 14</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5873b8"}}} -->
<p class="has-background" style="background-color:#5873b8">Level 15</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5873b8"}}} -->
<p class="has-background" style="background-color:#5873b8">Level 16</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5873b8"}}} -->
<p class="has-background" style="background-color:#5873b8">Level 17</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5873b8"}}} -->
<p class="has-background" style="background-color:#5873b8">Level 18</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5873b8"}}} -->
<p class="has-background" style="background-color:#5873b8">Level 19</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5873b8"}}} -->
<p class="has-background" style="background-color:#5873b8">Level 20</p>
<!-- /wp:paragraph --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group -->

<!-- wp:paragraph -->
<p></p>
<!-- /wp:paragraph --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group -->
```

</details>

2. Open the post in the app.
3. Observe that no crash is produced.
4. Scroll down the content.
5. Observe that starting at level 10, a warning message box is displayed within the block list.
6. Tap on the warning twice.
7. Observe that a bottom sheet is displayed explaining the problem.
8. Observe that the message doesn't mention the web editor.
9. Observe that only the "Ungroup block" option is presented.
10. Tap on the "Ungroup block" option.
11. Observe that a new block appears in the block list. This block should be the one that wasn't previously rendered due to being below level 10.
12. Observe that the warning message box is still present.
13. Tap on the warning message box and then the "Dismiss" option.
14. Observe that the bottom sheet closes.
15. Save the post.
16. Observe that no crash is produced.

#### Self-hosted site with Jetpack

1. In the web version of the editor, create a post with the following HTML in a self-hosted site with Jetpack:

<details><summary>Deeply nested block structure (20 levels)</summary>

```
<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#cb6767"}}} -->
<p class="has-background" style="background-color:#cb6767">Level 1 </p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#b89759"}}} -->
<p class="has-background" style="background-color:#b89759">Level 2</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#816531"}}} -->
<p class="has-background" style="background-color:#816531">Level 3</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#86c64d"}}} -->
<p class="has-background" style="background-color:#86c64d">Level 4</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#58b873"}}} -->
<p class="has-background" style="background-color:#58b873">Level 5</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#58b8b8"}}} -->
<p class="has-background" style="background-color:#58b8b8">Level 6</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5880b8"}}} -->
<p class="has-background" style="background-color:#5880b8">Level 7</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5870b8"}}} -->
<p class="has-background" style="background-color:#5870b8">Level 8</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5877b8"}}} -->
<p class="has-background" style="background-color:#5877b8">Level 9</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5873b8"}}} -->
<p class="has-background" style="background-color:#5873b8">Level 10</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5873b8"}}} -->
<p class="has-background" style="background-color:#5873b8">Level 11</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5873b8"}}} -->
<p class="has-background" style="background-color:#5873b8">Level 12</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5873b8"}}} -->
<p class="has-background" style="background-color:#5873b8">Level 13</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5873b8"}}} -->
<p class="has-background" style="background-color:#5873b8">Level 14</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5873b8"}}} -->
<p class="has-background" style="background-color:#5873b8">Level 15</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5873b8"}}} -->
<p class="has-background" style="background-color:#5873b8">Level 16</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5873b8"}}} -->
<p class="has-background" style="background-color:#5873b8">Level 17</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5873b8"}}} -->
<p class="has-background" style="background-color:#5873b8">Level 18</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5873b8"}}} -->
<p class="has-background" style="background-color:#5873b8">Level 19</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"constrained"}} -->
<div class="wp-block-group"><!-- wp:paragraph {"style":{"color":{"background":"#5873b8"}}} -->
<p class="has-background" style="background-color:#5873b8">Level 20</p>
<!-- /wp:paragraph --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group -->

<!-- wp:paragraph -->
<p></p>
<!-- /wp:paragraph --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group --></div>
<!-- /wp:group -->
```

</details>

2. Go to Jetpack/Settings and uncheck the option "Allow users to log in to this site using WordPress.com accounts" under the "WordPress.com login" section.
3. Open the post in the app.
4. Observe that no crash is produced.
5. Scroll down the content.
6. Observe that starting at level 10, a warning message box is displayed within the block list.
7. Tap on the warning twice.
8. Observe that a bottom sheet is displayed explaining the problem.
9. Observe that a note is included mentioning to allow `WordPress.com` login to edit using UBE.
10. Tap on the "Open Jetpack Security settings" option.
11. Enable "Allow `WordPress.com` login".
12. Tap on the warning message box again.
13. Observe that the option "Edit using web editor" is now presented.
14. Tap on the "Edit using web editor" option.
15. Observe that UBE (Unsupported Block Editor) opens with the content that is not rendered in the native editor.
16. Modify the content and tap on "Continue".
17. Preview the post and observe the post reflects recent modifications.
18. Back in the editor tap on the warning message box again.
19. Tap on the "Ungroup block" option.
20. Observe that a new block appears in the block list. This block should be the one that wasn't previously rendered due to being below level 10.
21. Observe that the warning message box is still present.
22. Tap on the warning message box and then the "Dismiss" option.
23. Observe that the bottom sheet closes.
24. Save the post.
25. Observe that no crash is produced.

#### Unsupported Block Editing

Since we are using UBE in the warning message box, it would be great to cover [some of the test cases of that feature](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/unsupported-block-editing.md).

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->
N/A

## Screenshots or screencast <!-- if applicable -->

| Light mode | Dark mode |
|--------|--------|
| <img width=300 src=https://github.com/WordPress/gutenberg/assets/14905380/36b3addc-8adb-4921-a359-a5f40b498f58> | <img width=300 src=https://github.com/WordPress/gutenberg/assets/14905380/3da91b38-522b-4f00-8381-fd1bbda1849d> |
| <img width=300 src=https://github.com/WordPress/gutenberg/assets/14905380/d3f68f1c-f62c-4574-a8a4-0c093cad2f47> | <img width=300 src=https://github.com/WordPress/gutenberg/assets/14905380/c6fc40ed-bf58-4bdf-86c2-45a40d9b1066> | 

| WPCOM site | Self-hosted site without Jetpack |
|--------|--------|
| <video src=https://github.com/WordPress/gutenberg/assets/14905380/afc68da1-d6e0-4558-b7fb-c051e181ceec> | <video src=https://github.com/WordPress/gutenberg/assets/14905380/aff8d09c-c947-4d7f-a45b-d5c226dcfb42> |

fluiddot committed:
Prevent rendering deeply nested block list

fluiddot committed:
Add `BlockFallbackWebVersion` component

fluiddot committed:
Add `WarningMaxDepthExceeded` component

fluiddot committed:
Add max depth warning in inner blocks component

fluiddot commented:
@osullivanchris I've shared some screenshots and video captures in the PR. Let me know if you have any feedback regarding design and copy, thanks 🙇 !

osullivanchris commented:
@fluiddot looks good. Nice job!

dcalhoun commented on packages/block-editor/src/components/inner-blocks/warning-max-depth-exceeded.native.js:
```diff
@@ -0,0 +1,84 @@
+
+	/* translators: Warning related to having blocks deeply nested. %d: The deepest nesting level. */
+	const descriptionFormat = __(
+		'Blocks with inner blocks that are nested deeper than %d levels might not be rendered properly in the native version of the editor. For this reason, we recommend flattening the content by ungrouping the block or editing the block using the web editor.'
```
Some tweaks for consideration. Feel free to ignore or cherry pick portions as you see fit. 

* Reduce sentence length. 
* Avoid technical terms layman users may not recognize, e.g. inner blocks, native. 
* Referencing web browser may be clearer than web editor. 

```suggestion
		'Blocks nested deeper than %d levels may not render properly in the mobile editor. For this reason, we recommend flattening the content by ungrouping the block or editing the block using your web browser.'
```

fluiddot replied:
> Referencing web browser may be clearer than web editor.

I explicitly mentioned the web editor because this is how it's referenced in the action button. My idea with the message is to explain what's happening but also give the option to the user to work around it.

<img src=https://github.com/WordPress/gutenberg/assets/14905380/4bcd80f5-81f3-40ca-bc66-9570c3f5b1dc width=300>

dcalhoun replied:
> I explicitly mentioned the web editor because this is how it's referenced in the action button.

Agreed this helps improve clarity. I did consider this as well. My suggestion to use "browser" was due to the UBE is not available to all users. The action button is conditional, but this message is currently static. 

I do not have a strong opinion on this particular matter. I do see value in "web editor" for the reasons you outlined.

fluiddot replied:
Ah, true. In case the web editor is not available to the user, we could display a different message mentioning the browser.

fluiddot replied:
Regarding the use of `web editor` and `web browser`, in https://github.com/WordPress/gutenberg/pull/54382/commits/87fbea640ef5863bc9e79318b7f958b4d836c5de I added both but each will be used depending on the availability of UBE.
- When UBE is available: Use web editor.
- WHen UBE is not available: Use web browser.

dcalhoun commented on packages/block-editor/src/components/inner-blocks/warning-max-depth-exceeded.native.js:
```diff
@@ -0,0 +1,84 @@
+// When creating a deep nested structure of inner blocks, the editor might exceed
+// this limit and crash. In order to avoid this, we set a maximum depth level where
+// we stop rendering blocks.
+export const MAX_NESTING_DEPTH = 10;
```
If we relocated this constant export to a different file, that would allow React Native to [Fast Refresh](https://reactnative.dev/docs/0.71/fast-refresh) the React component on changes without the need to perform a full reload or refresh other components.

fluiddot replied:
Good idea 👍 , I'll create a constants file and move this constant there.

fluiddot replied:
Solved in https://github.com/WordPress/gutenberg/pull/54382/commits/a4fc8725b17a34a56640ce1d801746f077f7ca88.

dcalhoun commented on packages/block-editor/src/components/inner-blocks/warning-max-depth-exceeded.native.js:
```diff
@@ -0,0 +1,84 @@
+					clientId={ clientId }
+					showSheet={ showDetails }
+					onCloseSheet={ () => setShowDetails( false ) }
+					title={ __( 'Block deeply nested' ) }
```
This suggestion is pretty much the same sentence, but for whatever reason this potential alternative sounds better to me. Possibly pure preference. Please feel free to ignore the suggestion. 
```suggestion
					title={ __( 'Deeply nested block' ) }
```

dcalhoun commented on packages/block-editor/src/components/inner-blocks/warning-max-depth-exceeded.native.js:
```diff
@@ -0,0 +1,84 @@
+	return (
+		<TouchableWithoutFeedback
+			disabled={ ! isSelected }
+			accessibilityLabel={ __( 'Warning message' ) }
+			accessibilityRole={ 'button' }
+			accessibilityHint={ __( 'Tap here to show more details.' ) }
```
I'm unsure as to why, but this button is announced as a "help" button. 

iOS: `Help button. Tap here to show help.`
Android: `Help button button. Tap here to show help. [Disabled|Double-tap to activate].` 

I believe the duplicative "button" term for Android is a preexisting issue. Do you know where the "help" term is originates?

dcalhoun replied:
From searching, the a11y description match the Unsupported Block Interface. However, I do not understand why they would be presented here. 

https://github.com/WordPress/gutenberg/blob/e028e9ada1c06cb8e31e9b848e492ad37422ba65/packages/block-library/src/missing/edit.native.js#L86-L99

fluiddot replied:
Yeah, that's unexpected. Have you experienced this when testing the installable build? If so, this is likely the same issue as the one mentioned [here](https://github.com/WordPress/gutenberg/pull/54382#discussion_r1327312563). I'll update the installable builds to reflect all changes, sorry for the inconvenience 🙇 .

dcalhoun replied:
Yeah, I have only used the installable build so far.

dcalhoun commented on packages/block-editor/src/components/block-fallback-web-version/style.native.scss:
```diff
@@ -0,0 +1,56 @@
+.sheet__container {
```
Nitpick: considering the BEM naming methodology, it feels odd to not name this merely `.sheet` IMO. 

Alternatively, given `sheet` is a reference to an implementation detail, I wonder if it makes more sense to name this `block-fallback` or similar.

dcalhoun commented on packages/block-editor/src/components/index.native.js:
```diff
@@ -91,6 +91,7 @@ export { default as __unstableEditorStyles } from './editor-styles';
 export { default as Inserter } from './inserter';
 export { useBlockProps } from './block-list/use-block-props';
 export { default as FloatingToolbar } from './floating-toolbar';
+export { default as BlockFallbackWebVersion } from './block-fallback-web-version';
```
Do we need to export this component or should it be considered internal/private?

fluiddot replied:
We don't need to export it in this PR because it's only used within the same package. However, my plan is to use this component also in [the Missing block](https://github.com/WordPress/gutenberg/blob/45f5608443d939dac1e21b3bedd0a2662992711b/packages/block-library/src/missing/edit.native.js#L295) (i.e. the one used to display unsupported blocks). This block is in a different package, so we'd need to export it for that case. That said, I agree that since it's not needed in these changes, we can remove the export and incorporate it in the future when needed 👍 .

fluiddot replied:
Solved in https://github.com/WordPress/gutenberg/pull/54382/commits/4ae212b19579900d790496b6ac0db30594551752.

fluiddot replied:
> [...] However, my plan is to use this component also in [the Missing block](https://github.com/WordPress/gutenberg/blob/45f5608443d939dac1e21b3bedd0a2662992711b/packages/block-library/src/missing/edit.native.js#L295) (i.e. the one used to display unsupported blocks). This block is in a different package, so we'd need to export it for that case. [...]

@dcalhoun In relation to this comment, I'd like to share that I opened [this PR](https://github.com/WordPress/gutenberg/pull/55133) to use the new `UnsupportedBlockDetails` component in the Missing block. Let me know if you could take a look, thanks 🙇 !

dcalhoun commented on packages/block-editor/src/components/block-fallback-web-version/index.native.js:
I am unsure where the name `web-version` originates. Is this because it conditionally offers the Unsupported Block Editor (UBE)? Is the intention to replace the UBE's bottom sheet with this one? IMHO this abstraction/location feels premature unless we have a specific place we intend to reuse this component. 

WDYT?

fluiddot replied:
> I am unsure where the name `web-version` originates. Is this because it conditionally offers the Unsupported Block Editor (UBE)?

Yes, I used that name because it opens the web version of the editor. However, I agree it's not the best naming 😅, so I'm open to suggestions. Maybe something related to UBE like: `UnsupportedBlockEditorBottomSheet`.

> Is the intention to replace the UBE's bottom sheet with this one? IMHO this abstraction/location feels premature unless we have a specific place we intend to reuse this component.

Yep, that's why I extracted the logic to a separate component. I mentioned this in the PR's description:

> Additionally, I extracted the UI elements used in [the "missing" block](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/missing/edit.native.js) (i.e. the block used to render unsupported blocks) into a separate component (BlockFallbackWebVersion). In a separate PR, I'll update that block to use this new component.

dcalhoun replied:
Aha, apologies for overlooking the note in the description. 🙇🏻‍♂️ 

I think `UnsupportedBlockEditorBottomSheet` or something like `UnsupportedBlockDetails` (given the editor is not always available) would work well. 👍🏻

fluiddot replied:
I originally avoided using the word `Unsupported`, as in this case, it's not strictly an unsupported block but an editor's limitation. In any case, I can't think of a more accurate name 😅. `UnsupportedBlockDetails` is short and self-explaning.

fluiddot replied:
Solved in https://github.com/WordPress/gutenberg/pull/54382/commits/053a2c83a5726f175448c6a2f41082806aae1f3b.

dcalhoun commented on packages/block-editor/src/components/block-fallback-web-version/index.native.js:
```diff
@@ -0,0 +1,207 @@
+		( isUnsupportedBlockEditorSupported ||
+			canEnableUnsupportedBlockEditor ) &&
+		isEditableInUnsupportedBlockEditor;
+	const actions = [
+		...[
+			canUseWebEditor && {
+				label: actionButtonLabel || __( 'Edit using web editor' ),
+				onPress: requestFallback,
+			},
+		],
+		...customActions,
+	].filter( Boolean );
```
I noticed in the prototype builds that sites that do not have access to the Unsupported Block Editor display an empty action. 

| iOS | Android |
| - | - |
| ![empty-action-ios](https://github.com/WordPress/gutenberg/assets/438664/850cd043-d0fb-47aa-8ec6-44801f374e09) | ![empty-action-android](https://github.com/WordPress/gutenberg/assets/438664/746fffcf-9b20-4f4f-aa11-5daa557c2ae2) |

fluiddot replied:
Ah, that's interesting. I bumped into this issue while testing before setting the PR as ready to review, but I thought I had fixed it 🤔. I'll take a look in case I overlooked anything in the changes. Thanks!

dcalhoun replied:
Maybe the prototype builds are merely using an outdated Gutenberg ref? I did not check.

fluiddot replied:
Ah, that's likely the case as I haven't recently updated the host app PRs. I'll update them asap.

dcalhoun commented on packages/block-editor/src/components/block-fallback-web-version/index.native.js:
```diff
@@ -0,0 +1,207 @@
+					setSendFallbackMessage( false );
+				} else if ( sendButtonPressMessage ) {
+					this.timeout = setTimeout( () => {
+						sendActionButtonPressedAction(
+							actionButtons.missingBlockAlertActionButton
+						);
```
When is this logic utilized? I'm struggling to follow what the outcome of this logic. 

My perception is that this is preexisting code copied from the Unsupported Block Editor sheet.

fluiddot replied:
Good question, I agree the action's name is not very descriptive 😅. This action is used when UBE can be used to edit the block but it's disabled for some reason. Usually, this happens in self-hosted sites connected to Jetpack that don't have the `Login with WordPress.com` option enabled. In this case, the user is prompted to enable this option and re-open the web editor again.

I checked the UBE flow in a self-hosted site with Jetpack and noticed we'd need to make some adjustments. I'll work a on fix for this 🔧 .

fluiddot replied:
Here is an example of the flow for the above case:


https://github.com/WordPress/gutenberg/assets/14905380/39950ac0-804e-4919-b8b2-53b19604eaa9

dcalhoun reviewed (commented):
Well done! Thank you for identifying a useful resolution to this crash. 

I left several comments for consideration. Additionally, I noted that one can change the background color, text color, and alignment of these blocks. Only the first of those three results in a visual change. Is this expected? Should we disable those controls? Is it confusing? Does it matter? I'm not sure. 🤔 

<img alt="block-background-color-and-alignment" src="https://github.com/WordPress/gutenberg/assets/438664/3b148a77-673d-4edd-969b-2567ac53d9cf" width="200" />

fluiddot committed:
Update title of max depth exceeded warning

Co-authored-by: David Calhoun <github@davidcalhoun.me>

fluiddot commented:
> Well done! Thank you for identifying a useful resolution to this crash.
> 
> I left several comments for consideration. Additionally, I noted that one can change the background color, text color, and alignment of these blocks. Only the first of those three results in a visual change. Is this expected? Should we disable those controls? Is it confusing? Does it matter? I'm not sure. 🤔
> 
> <img alt="block-background-color-and-alignment" width="200" src="https://user-images.githubusercontent.com/438664/268093799-3b148a77-673d-4edd-969b-2567ac53d9cf.jpg">

Good question. The warning limit is displayed within the block's content, so the rest of the block settings, like the toolbar buttons, are still present. I don't have a strong opinion on whether we should disable the settings. However, if we'd like to allow the user to still edit the block, even though some of them won't have a visual change, maybe it makes sense to keep them. Besides, those changes would be reflected when editing the block via UBE or in case the block is ungrouped.


https://github.com/WordPress/gutenberg/assets/14905380/4c51e598-5e61-4064-a15f-b1bfb5e59f6b

**UPDATE:**

> [...] Besides, those changes would be reflected when editing the block via UBE or in case the block is ungrouped.

Seems that changes to block settings are not propagated (at least the text color) when ungrouping a block. I checked this behavior in the web version and it can also be reproduced there.

dcalhoun commented:
> I don't have a strong opinion on whether we should disable the settings. However, if we'd like to allow the user to still edit the block, even though some of them won't have a visual change, maybe it makes sense to keep them.

Given this is somewhat of an edge case within an edge case, I feel comfortable leaving the implementation as-is. It could cause some confusion, but it is not destructive in any way.

fluiddot committed:
Update description of max depth exceeded warning

fluiddot committed:
Add constants file to `InnerBlocks` component

`MAX_NESTING_DEPTH` has been moved to the constants file.

fluiddot committed:
Revert `BlockFallbackWebVersion` export

fluiddot committed:
Rename `BlockFallbackWebVersion` component to `UnsupportedBlockDetails`

fluiddot committed:
Rename `UnsupportedBlockDetails` styles

fluiddot committed:
Update default web editor action label via WP hooks

fluiddot committed:
Allow inserting extra notes to description via WP hooks

fluiddot committed:
Extract UBE settings fetch to a hook

fluiddot committed:
Update description of max depth exceeded warning based on UBE support

fluiddot commented:
@dcalhoun I addressed the feedback from the review and generated new installable builds. Let me know if you could take another look.

- [WordPress-Android](https://github.com/wordpress-mobile/WordPress-Android/pull/19163#issuecomment-1716034886)
- [WordPress-iOS](https://github.com/wordpress-mobile/WordPress-iOS/pull/21553#issuecomment-1716102889)

I also added two new WP hooks to allow GBM edit the messaging based on the UBE's availability (https://github.com/WordPress/gutenberg/pull/54382/commits/6e642708a17080d7c522f255d3a6f8355f769703, https://github.com/WordPress/gutenberg/pull/54382/commits/167418fc621935067d6652e20b0303ae2ed248fc).

| Simple site | Self-hosted without Jetpack | Self-hosted with Jetpack |
|--------|--------|--------|
| <img src=https://github.com/WordPress/gutenberg/assets/14905380/8a466246-6d68-469b-866f-67a3b4ac982d> | <img src=https://github.com/WordPress/gutenberg/assets/14905380/b9e6d9dc-27d9-4b2e-8c3a-13f6026bf6e4> | <img src=https://github.com/WordPress/gutenberg/assets/14905380/89c57530-fec8-4846-b64f-c92f5b7c3a43> |

dcalhoun reviewed (approved):
🚀 I verified the test plan succeeds. Thank you for improving this experience.

dcalhoun commented:
We should likely add a release note.

fluiddot commented:
> We should likely add a release note.

Good point. I'll add an entry to the changelog and release notes before merging the PR.

fluiddot committed:
Merge branch 'trunk' into rnmobile/fix/call-stack-size-exceeded-limit-solution

fluiddot committed:
Update `react-native-editor` changelog

fluiddot committed:
Merge branch 'trunk' into rnmobile/fix/call-stack-size-exceeded-limit-solution
